compiler/rustc_codegen_ssa/src/sir.rs

```diff
@@ -890,3 +890,136 @@ impl SirTypes {
+                                    panic!("Error reading dwarf information. Expected type 'Addr'.")
+                                }
+                            };
+                            // We can not accurately insert labels at the beginning of functions,
+                            // because the label is offset by the function headers. We thus simply
+                            // remember the subprogram's address so we can later assign it to the first
+                            // block (ending with '_0') of this subprogram.
+                            subaddr = Some(addr);
+                        }
+                    }
+                } else if entry.tag() == gimli::DW_TAG_label {
+                    if let Some(name) = entry.attr_value(gimli::DW_AT_name)? {
+                        if let Some(es) = name.string_value(&dwarf.debug_str) {
+                            let s = es.to_string()?;
+                            if s.starts_with("__YK_") {
+                                if let Some(lowpc) = entry.attr_value(gimli::DW_AT_low_pc)? {
+                                    let addr = match lowpc {
+                                        gimli::AttributeValue::Addr(v) => v as u64,
+                                        _ => panic!(
+                                            "Error reading dwarf information. Expected type 'Addr'."
+                                        ),
+                                    };
+                                    if subaddr.is_some() && s.ends_with("_0") {
+                                        // This is the first block of the subprogram. Assign its label
+                                        // to the subprogram's address.
+                                        let (fsym, bb) = split_symbol(s);
+                                        labels.push(ykpack::SirLabel {
+                                            off: usize::try_from(subaddr.unwrap()).unwrap(),
+                                            symbol_name: fsym,
+                                            bb,
+                                        });
+                                        subaddr = None;
+                                    } else {
+                                        let (fsym, bb) = split_symbol(s);
+                                        labels.push(ykpack::SirLabel {
+                                            off: usize::try_from(addr).unwrap(),
+                                            symbol_name: fsym,
+                                            bb,
+                                        });
+                                    }
+                                } else {
+                                    // Ignore labels that have no address.
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        labels.sort_by_key(|l| l.off);
```

Comment:
Is this step necessary or is it a bit of OCD? :)

Reply:
If we maintain the list in order, we can do an early bail out optimisation at runtime.

[We currently have to search a list of address ranges. This would be better as an interval tree, but that's another story].

Reply:
Let's document that in a function doc comment: that's a *very* important invariant that the function is required to maintain.